### DISCUSSION TITLE
Embbed support for geographic types.

### DISCUSSION THREAD

**Discussion by oreilles** [directus/directus#5193] [Feature Requests]
In the current map layout/interface implementation, users have to choose themselves the format, projection, type (point, polygon) in order to display/edit them properly. It is very much error prone as there is no binding between the interface and layout options, and doesn't offer any protection towards data integrity because interface options could change anytime, leading to incompatible geographic types inside one column.

Ideally, all those options should be defined at the row creation and locked - or infered from the database (like it is the case for say string size or floating point precisions).
Thus the need for built in geographic types.

We could either add a `geometry` option to the field types, or a new section with one different type for `point`, `polygon`, `linestring`. In both cases, the *real* storage type (GeoJSON, PostGIS, EWKT...) would have to be defined in that field's schema options, alongside with type, multiple_allowed and projection.

But then I'm also starting to wonder if we should only support GeoJSON and PostGIS for Postgres users. With help of [knex-postgis](https://github.com/jfgodoy/knex-postgis), we could read/write postgis data as reprojected GeoJSON directly, and we then wouldn't need parsing / serializing / reprojecting altogether. (GeoJSON data is required to be in WGS84 per specifications). It would vastly simplify the current code and improve performance.

Any thoughts on all of this ?

**Comment by rijkvanzanten** [comment]
_Disclaimer: I'm (clearly) no expert on the specifics when it comes to saving / working with location data.._

The main problem I'm seeing with this is the availability per database vendor 🤔 I'd rather not go the route where some features are only available in certain databases. It seems like GeoJSON is a type we could check for in the API, save in every database (with fallbacks to JSON and TEXT where needed), and progressively enhance as datatype for the vendors that support it natively. 

Is there a single (or maybe top 2-3) data standard that's widespread enough to use as a spec for "location type in Directus" (similar to how we're using ISO8601 for dates), or is it too usage-based / non-standard to enforce such a thing?

**Reply by oreilles** [reply]
GeoJSON (json) and (E)WKT (text) are the most common standardised format. 

> I'd rather not go the route where some features are only available in certain databases. 

Totally get that. 👍 Supporting PostGIS read/write wouldn't add any feature though, to an end-user it could even be completely transparent. There would really be only two/three function used: `ST_AsGeoJSON`, `ST_GeomFromGeoJSON` and `ST_Transform`. Those functions are exposed by PostGIS and Spatialite(basically PostGIS for SQLite), but they also are built into MySQL and MariaDB, which both have native geographic data types! We could just check if these are available and fallback to a json field if they aren't. 

Postgres+PostGIS is one if not the most widespread spatial database solution, having builtin support for it would be a real advantage for Directus, whose one of the main selling point for me was the ability to be plugged in preexisting database. There basically don't exist any geospatial app or CMS on the market that isn't a full-blown GIS. 🙂

**Reply by rijkvanzanten** [reply]
Is it safe to say that we could consider GeoJSON the standardized IO for use in the API, or would that be too limited? 

Also semi-unrelated, do you have some good sources for me to read up on? I'd love to know more about all this, but it's a little overwhelming with all the different types and providers 😄

**Comment by oreilles** [comment]
> Is it safe to say that we could consider GeoJSON the standardized IO for use in the API, or would that be too limited?

Yes, I think that's the way to go. The only thing that would imply is the use of the functions I talked about for reading/writing to native geometry columns, as those are binary format and not JSON.

About some resource for you to read up, not sure where to guide you first!  If you just want a foot in, maybe OpenLayers and Mapbox-gl-js documentations/tutorials could be good start..
I mean geospatial data science is a big thing, between coordinates reference systems , serialization formats, data exchange protocols, visualisation libraries, database solutions... But ultimately we'll just do simple stuff here. Get some GeoJSON from the database, feed it to maplibre-gl and vice-versa. 👍

**Reply by rijkvanzanten** [reply]
> I mean geospatial data science is a big thing

Overwhelmingly so! 😅

**Reply by benhaynes** [reply]
@Oreilles — are you on [our Discord](https://directus.chat)? Or is there an email where I can reach you? I'd love to discuss a few things with you outside the public eye... 👁️

**Reply by oreilles** [reply]
@benhaynes  I just answered back your pm on Discord 👍

**Reply by cj** [reply]
We use PostGIS in a project (with Rails), this is a great article https://pganalyze.com/blog/postgis-rails-geocoder

I am seriously considering using Directus as an admin now it has switched to node, and this is something I would definitely be implementing as a type.

**Reply by oreilles** [reply]
PostGIS is already supported in the current implementation so if you want to try it out with your database you're good to go 👍 Otherwise, I don't think we'd want a type for each input/output format, but rather have classic geographic type (Point, Polygon...) and a compatibility layer to handle the conversion from native database geometries to geojson and vice-versa on the database side when they are supported.

**Reply by iyobo** [reply]
There is now a MAP field type as of the latest release of diretus where this has been implemented.

**Comment by jaads** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
